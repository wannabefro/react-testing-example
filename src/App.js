import React from 'react';
import TodosList from './TodosList';
import { generateId } from './utils';

export const ENTER_KEY = 13;

class App extends React.Component {
  state = {
    newTodo: '',
    todos: [],
  };

  onChange = e => {
    this.setState({ newTodo: e.target.value });
  }

  onKeyDown = e => {
    if (e.keyCode !== ENTER_KEY) return;
    e.preventDefault();

    const { newTodo, todos } = this.state;
    const title = newTodo.trim();

    if (title) {
      this.setState({ todos: [...todos, {
        id: generateId(),
        title,
      }], newTodo: ''});
    }
  }

  render() {
    const { todos, newTodo } = this.state;
    return (
      <div>
        <header className='header'>
          <h1>Todos</h1>
        </header>
        <input
          className='new-todo'
          value={newTodo}
          placeholder='What needs to be done?'
          onChange={this.onChange}
          onKeyDown={this.onKeyDown}
        />
        <TodosList todos={todos} />
      </div>
    );
  }
}

export default App;
