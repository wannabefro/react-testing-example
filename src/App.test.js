import React from 'react';
import { render, fireEvent } from 'react-testing-library';
import { shallow } from 'enzyme';
import * as utils from './utils';
import App, { ENTER_KEY } from './App';

test('creating a new todo', () => {
  const { getByPlaceholderText, getByTestId, asFragment } = render(<App />);
  const input = getByPlaceholderText('What needs to be done?');

  fireEvent.change(input, { target: { value: 'React testing' } });
  fireEvent.keyDown(input, { keyCode: 13 });

  expect(getByTestId('todosList')).toHaveTextContent('React testing');
  expect(asFragment()).toMatchSnapshot();
});

describe('App', () => {
  const value = 'Bob';
  const id = 123;

  jest.mock('./TodosList');
  jest.mock('./utils', () => ({
    generateId: jest.fn(() => 123),
  }));
  let wrapper;

  beforeEach(() => {
    utils.generateId = jest.fn(() => id);
    wrapper = shallow(<App />);
    wrapper.find('input').simulate('change', {
      target: { value },
    });
  });

  describe('onKeyDown', () => {
    describe('when the keycode is 13', () => {
      beforeEach(() => {
        wrapper.find('input').simulate('keyDown', { keyCode: ENTER_KEY, preventDefault: jest.fn()});
      });

      it('creates a new todo', () => {
        const todos = wrapper.find('TodosList').props().todos;
        expect(todos.length).toEqual(1);

        const [todo] = todos;
        expect(todo.title).toEqual(value);
        expect(todo.id).toEqual(123);
      });

      it('resets the value of the input', () => {
        expect(wrapper.find('input').props().value).toEqual('');
      });
    });

    describe('when the keycode is not 13', () => {
      it('does nothing', () => {
        expect(wrapper.find('input').props().value).toEqual(value);
        expect(wrapper.find('TodosList').props().todos.length).toEqual(0);
      });
    });
  });
});
