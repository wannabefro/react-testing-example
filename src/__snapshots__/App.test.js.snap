// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creating a new todo 1`] = `
<DocumentFragment>
  <div>
    <header
      class="header"
    >
      <h1>
        Todos
      </h1>
    </header>
    <input
      class="new-todo"
      placeholder="What needs to be done?"
      value=""
    />
    <section
      class="main"
    >
      <ul
        class="todo-list"
      >
        <li>
          <label>
            React testing
          </label>
        </li>
      </ul>
    </section>
  </div>
</DocumentFragment>
`;
